### ISSUE TITLE
Equality between u8 is rejected

### ISSUE THREAD

**Issue by yannickmoy** [xldenis/creusot#304]
The following code is rejected:

```rust
extern crate creusot_contracts;

use creusot_contracts::*;

#[predicate]
fn match_at (needle: &[u8], haystack: &[u8], pos: usize) -> bool {
  pearlite! { pos >= 0usize && pos <= haystack.len() - needle.len()
    && forall<i: usize>
        i >= 0usize && i < needle.len() ==> needle[i] == haystack[pos + i]
  }
}
```

on error ``error[E0599]: no method named `log_eq` found for type `u8` in the current scope``

I tried changing the `==` into logical equality `===` inside the quantification, but then I get another error ``thread 'rustc' panicked at 'not yet implemented: lower_expr: Index { lhs: e28, index: e30 }', creusot/src/translation/specification/typing.rs:271:19``

Is there a way to write the function `match_at` so that Creusot accepts it? thanks

**Comment by sarsko** [comment]
If you can accept going from arrays to vectors, then I believe the following should be what you want.

```rust
extern crate creusot_contracts;
use creusot_contracts::*;
use creusot_contracts::std::*;

#[predicate]
fn match_at (needle: &Vec<u8>, haystack: &Vec<u8>, pos: usize) -> bool {
  pearlite! { @pos >= 0 && @pos <= (@haystack).len() - (@needle).len()
    && forall<i: Int>
        0 <= i && i < (@needle).len() ==> (@needle)[i] === (@haystack)[@pos + i]
  }
}
```

**Comment by yannickmoy** [comment]
thanks @sarsko, just changing `[u8]` into `Vec<u8>` makes it acceptable for Creusot!
Is there any reason why you prefer to express the predicate all in terms of models rather than in terms of `u8` and `usize`? Both are accepted by Creusot.

**Comment by sarsko** [comment]
Are you getting it to load in Why3 with just a swap to `Vec<u8>`? For me that results in the Why3 error "unbound function or predicate symbol ..."

**Comment by xldenis** [comment]
> I tried changing the == into logical equality === inside the quantification, but then I get another error thread 'rustc' panicked at 'not yet implemented: lower_expr: Index { lhs: e28, index: e30 }', creusot/src/translation/specification/typing.rs:271:19

Oh, that sucks, it happens because slices are desugared differently from normal indexing, I'll have to fix that manually. 

> Is there any reason why you prefer to express the predicate all in terms of models rather than in terms of u8 and usize? Both are accepted by Creusot.

As @sarsko said, this would be rejected later on by why3 as you will be using the _program_ functions for indexing your vector. In @sarsko's version of your code it uses `@needle` which provides the model as a `Seq`. Finding a way to properly seperate the program and logical namespaces in Rust is something which remains to be done.

**Comment by yannickmoy** [comment]
as @sarsko said, I'm getting `unbound type symbol 'array'` later. I'll use the model version!

**Comment by xldenis** [comment]
I fixed this a few weeks ago in #314
